proposals/accepted/051-ghc-base-libraries.rst

```diff
@@ -0,0 +1,316 @@
+
+In the past we have relied on best efforts; but with a bunch of volunteers, mistakes will be made.
+And mistakes can lead to a loss of trust.
+
+The solution is obvious: we need to automate.
+We therefore propose the following, as part of CI:
+
+1. Compile a good chunk of Hackage (around 500 packages) against base.
+   We already do this, and it is a huge help in reassuring ourselves that a change does not lead to accidental breakage.
+
+2. Test if any of the types (incl their kinds), functions (incl their types) and instances exposed by the ``base`` API are accidentally changed by a commit.
+   This is definitely going to happen, soon: @bgamari already has a prototype.
+
+3. Run the test suite of those packages that have a testsuite that
+
+   (a) is usable (e.g. that doesn't take too long to run),
+
+   (b) does not have dependencies that are outside the set mentioned in point (1), and
+
+   (c) passes before the change to GHC/``base``.
+
+   This checks semantics as well as types.
+
+4. Running the performance test suite of some carefully chosen packages.
+   This checks for performance regressions.
+   Similar to (3), except that perf suites are less common and often more expensive to run.
+
+5. Develop a new suite of performance tests, specifically for base.
+   This is quite open-ended; it is not clear what would be desirable, or how much it would cost.
+
+Some modules in ``ghc-internals`` will very directly affect exports of ``base`` (e.g via shim).
+These modules could be identified, via the existing ``CODEOWNERS`` mechanism, to ping CLC on any commit to those modules.
+This list could be selective, or include all of ``ghc-internals``, at CLC's preference.
+
+Some of these are cheap to do; others are less so.
+Fortunately the HF seems willing to help.
+
+*But whatever we do here will be a step forward* from our current, unsatisfactory situation.
+Moreover, they will help with CI for changes to GHC itself! (It is rather *more* likely that a commit to GHC's simplifier will cause a perf regression in some package, than a commit to ``ghc-internals``.)
+
+Discussion
+==========
+
+GHC Proposals process
+---------------------
+
+Some GHC proposals (a minority) directly affect the existing API of ``base``, and are not simply additions that can be exposed in ``ghc-experimental``.
+It is unproductive for the GHC Steering Committee to have a long discussion, accept the proposals, and only *then* involve the CLC.
+
+We propose that:
+
+- A GHC Proposal should advertise, in a separate section, that it makes changes to the ``base`` API, and what those changes are.
+
+- If there are any such changes, the author (and shepherd) should explicitly invite the CLC to participate in the discussion about the proposal.
+  The CLC will devote some effort to participating, and subsequently hold a non-binding vote.
```

Comment:
The CLC will devote some effort to participating and, in the case of changes to `base`, will subsequently hold a non-binding vote.